File: Tree Data Structure/flattenBinaryTreeToLinkedList.py
class TreeNode:
    def __init__(self, val):
        self.val = val
        self.left = None
        self.right = None


class Solution:
    # @param A : root node of tree
    # @return the root node in the tree
    def __init__(self):
        self.dummyNode = TreeNode(0)
        self.temp = self.dummyNode

    def preOrder(self, root):
        if root is None:
            return
        self.dummyNode.right = TreeNode(root.val)
        self.dummyNode = self.dummyNode.right
        # print(root.val)
        self.preOrder(root.left)
        self.preOrder(root.right)
        root = None

    def flatten(self, A):

        self.preOrder(A)

        return self.temp.right

        # def preOrder(self, root, arr):
        #     if root is None:
        #         return
        #     arr.append(root.val)
        #     self.preOrder(root.left, arr)
        #     self.preOrder(root.right, arr)

        # def flatten(self, A):
        #     arr = []
        #     self.preOrder(A, arr)
        #     print(arr)
        #     dummyNode = TreeNode(0)
        #     temp = dummyNode

        #     for i in range(len(arr)):
        #         temp.right = TreeNode(arr[i])
        #         temp = temp.right

        # return dummyNode.right

    def print(self, node):
        while node is not None:
            print(node.val)
            node = node.right


node = TreeNode(1)
node.left = TreeNode(2)
node.left.left = TreeNode(3)
node.left.right = TreeNode(4)
node.right = TreeNode(5)
node.right.right = TreeNode(6)

s = Solution()
print(s.print(s.flatten(node)))

File: FacebookInterviewQuestions/flattenBinaryTreeToLinkedList.js
var flatten = function(root) {
  let stack = [];
  stack.push(root);
  while (stack.length > 0) {
    let current = stack.pop();
    if (current === null) continue;
    if (current.right !== null) {
      stack.push(current.right);
    }
    if (current.left !== null) {
      stack.push(current.left);
    }

    if (stack.length > 0) {
      let idx = stack.length - 1;
      current.right = stack[idx];
    }

    current.left = null;
  }
};
class TreeNode {
  constructor(val) {
    this.val = val;
    this.left = null;
    this.right = null;
  }
}

let node = new TreeNode(1);
node.left = new TreeNode(2);
node.left.left = new TreeNode(3);
node.left.right = new TreeNode(4);
node.right = new TreeNode(5);
node.right.right = new TreeNode(6);

function print(node) {
  while (node) {
    console.log(node.val);
    node = node.right;
  }
}
flatten(node);
print(node);

/**
 * public void flatten(TreeNode root) {
    flatten(root,null);
}
private TreeNode flatten(TreeNode root, TreeNode pre) {
    if(root==null) return pre;
    pre=flatten(root.right,pre);    
    pre=flatten(root.left,pre);
    root.right=pre;
    root.left=null;
    pre=root;
    return pre;
}

private TreeNode prev = null;

public void flatten(TreeNode root) {
    if (root == null)
        return;
    flatten(root.right);
    flatten(root.left);
    root.right = prev;
    root.left = null;
    prev = root;
}
 */
